### PR TITLE
David Atkins

### PR DESCRIPTION
**Pull request by d-atkins**
- Got through user story 4 and extension
- I felt like I did okay, still confused on what logic belongs where in the MVC

### CODE REVIEW

File: app/controllers/project_contestants_controller.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class ProjectContestantsController < ApplicationController
+  def update
+    project = Project.find(params[:project_id])
+    contestant = Contestant.where(id: params[:contestant_id])
+    project.contestants << contestant
+    redirect_back(fallback_location: '/contestants')
+  end
```
Comment (megstang) [inline app/controllers/project_contestants_controller.rb]: I would have this in a create action since you're creating a new entry onto this table by adding a contestant to a project. Great attempt at this especially knowing which controller to put this in and knowing that you had to add to the table and not create a brand new contestant or a brand new project. I know we went over this in the lesson, but here is how I would handle this action: 

```ruby 
def create 
ProjectContestant.create(project_id: params[:project_id],contestant_id:params[:contestant_id])
redirect_to "/projects/#{params[:project_id]}" # this should be redirecting to the projects show page according to the story. 
end 

```

File: app/models/project.rb [hunk 2]
```diff
@@ -1,4 +1,14 @@
...
+  has_many :contestant_projects
+  has_many :contestants, through: :contestant_projects
+
+  def number_of_contestants
+    contestants.count
+  end
+
+  def average_contestant_experience
+    contestants.sum(:years_of_experience).to_f / number_of_contestants
+  end
```
Comment (megstang) [inline app/models/project.rb]: great job with these... you can use the average method in activerecord next time 😉

File: app/views/contestants/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<% @contestants.each do |contestant| %>
+  <section id="contestant-<%= contestant.id %>">
+    <p><%= contestant.name %></p>
+    <p>Projects:
+      <% contestant.projects.each do |project| %>
+        <%= "#{project.name}," if project !=  contestant.projects.last %>
+        <%= "#{project.name}" if project ==  contestant.projects.last %>
```
Comment (megstang) [inline app/views/contestants/index.html.erb]: I like this creativity here

File: app/views/projects/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+<p><%= @project.name %></p>
+<p>Material: <%= @project.material %></p>
+<p>Challenge Theme: <%= @project.challenge.theme %></p>
+<p>Number of Contestants: <%= @project.number_of_contestants %></p>
+<% unless @project.average_contestant_experience.nan? %>
+  <p>Average Contestant Experience: <%= @project.average_contestant_experience.round(2) %></p>
+<% end %>
+<br>
+<%= form_tag("/projects/#{@project.id}", method: :patch) do  %>
```
Comment (megstang) [inline app/views/projects/show.html.erb]: since you're routing this to the projectcontestants controller, have your route reflect that `/project_contestants/:project_id`

File: app/views/projects/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<p><%= @project.name %></p>
+<p>Material: <%= @project.material %></p>
+<p>Challenge Theme: <%= @project.challenge.theme %></p>
+<p>Number of Contestants: <%= @project.number_of_contestants %></p>
+<% unless @project.average_contestant_experience.nan? %>
```
Comment (megstang) [inline app/views/projects/show.html.erb]: 💯 love this

File: config/routes.rb [hunk 6]
```diff
@@ -1,3 +1,6 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  get '/projects/:project_id', to: 'projects#show'
+  patch '/projects/:project_id', to: 'project_contestants#update'
```
Comment (megstang) [inline config/routes.rb]: again, just to reiterate, I'd make this: 

`post '/project_contestants/:project_id' , to: 'project_contestants#create'`

File: spec/features/contestants/index_spec.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        expect(page).to_not have_content(project3.name)
+      end
+
+      within("#contestant-#{contestant2.id}") do
+        expect(page).to have_content(contestant2.name)
+        expect(page).to have_content("Projects: #{project3.name}")
+        expect(page).to_not have_content(contestant1.name)
+        expect(page).to_not have_content(project1.name)
+        expect(page).to_not have_content(project2.name)
+      end
```
Comment (megstang) [inline spec/features/contestants/index_spec.rb]: great use of within here 👍

File: spec/features/projects/show_spec.rb [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+      project1 = Project.create!(name: "Litfit", material: "Lamp Shade", challenge: challenge1)
+
+      visit "/projects/#{project1.id}"
+
+      expect(page).to have_content("Litfit")
+      expect(page).to have_content("Material: Lamp Shade")
+      expect(page).to have_content("Challenge Theme: Apartment Furnishings")
+      expect(page).to have_content("Number of Contestants: 0")
+      expect(page).to_not have_content("Average Contestant Experience:")
+    end
```
Comment (megstang) [inline spec/features/projects/show_spec.rb]: good edge case test

File: spec/models/project_spec.rb [hunk 9]
```diff
@@ -12,4 +12,30 @@
...
+      challenge1 = Challenge.create!(theme: "Apartment Furnishings", project_budget: 50)
+      project1 = Project.create!(name: "Litfit", material: "Lamp Shade", challenge: challenge1)
+      contestant1 = Contestant.create!(name: "Kentaro Kameyama", age: 32, hometown: "Denver, CO", years_of_experience: 10)
+      contestant2 = Contestant.create!(name: "Jay McCarroll", age: 37, hometown: "Austin, TX", years_of_experience: 15)
+
+      project1.contestants << contestant1
+      project1.contestants << contestant2
+
+      expect(project1.average_contestant_experience).to eq(12.5)
+    end
```
Comment (megstang) [inline spec/models/project_spec.rb]: nice job remembering to test these methods
